1-Draft/RFC00XX-Update-Error-View.md

```diff
@@ -0,0 +1,204 @@
+---
+RFC: Update $ErrorView with simplified views for Error Messages
+Author: Jason Helmick
+Status: Draft
+Version: 0.0.0
+Area: PowerShell
+Comments Due: 10/31/2019
+---
+
+# Update $ErrorView with simplified views for Error Messages
+
+When an error occurs in PowerShell, the customers on-screen error message experience currently
+provides a level of detail that obscures the exception message from being recognized and read by
+new or occasional PowerShell users. The addition of a simplified error view will improve both
+comprehension and troubleshooting experience. A new cmdlet 'Resolve-ErrorRecord' will provide
+complete detailed view of the fully qualified error when desired.
+
+## Motivation
+
+The on-screen experience, when receiving an error message,
+is controlled through the views NormalView (the default) and CategoryView. These are user selectable
+through the preference variable $ErrorView.
+This RFC describes Changing '$ErrorView to an enumeration and adding three additional views
+to improve readability,'Message', 'Analytic', and 'Dynamic'.
+
+- Message - provides a concise error message suitable for new or occasional PowerShell users.
+- Analytic - provides a refactored form of NormalView
+- Dynamic - provides conditional switching between Message and Analytic.
+
+A comprehensive detailed view of the fully qualified error, including inner exceptions,
+will be provided by the `Resolve-ErrorRecord` cmdlet.
+
+'$ErrorView' shall contain the original views for backward compatibility and renamed
+to remove 'view' from the name. The view list is as follows:
+
+- Normal
+- Category
+- Message
+- Analytic
+- Dynamic
+
+## Specification
+
+The proposal is to add three new views to help improve error message comprehension
+based on user needs. For in-depth troubleshooting, a new cmdlet
+Resolve-ErrorRecord to provide detailed error information.
+
+__Key Design Considerations__
+
+1. To reduce confusion and improve debugging success for new and occasional users,
+error messages should call WriteErrorLine to produce a simplified message using view 'Message'
+and include the preface word “ERROR:”
+
```

Comment:
Why does "ERROR:" need to be included in the message? WriteErrorLine can prefix it. Especially since you are going to play with the color on those six characters, it seems that that actually complicates the situation.

Reply:
@essentialexch Great point!  In some of my testing I have it removed and I actually like it without the ERROR:.  Interested how others feel about having ERROR or not having ERROR?

Reply:
In the current code, if the error is an Exception, it prefixes with `Exception:`.  If it has InvocationInfo.MyCommand, it uses that.  If it has InvocationName, it uses that.  If it has CategoryInfo.Category, it uses that.  If it has CategoryInfo.Reason it'll use that.  Only if none of those exist does it actually use `Error:`.

Reply:
It is better to have a consistent, searchable word to allow quickly finding errors from a long stream of output. 

For example, I kick off run a long-running script that will loop through a series of actions such as compressing and moving files. I want it to be non-terminating errors so that it skips the file and keeps going. Now I come back to review the output, I want to search to see what errors occurred so I can fix specific issues with those files. How do I consistently know how to find the error output?

`Resolve-ErrorRecord` and `Error[x]` stack helps a bit if done interactively, but what if I was redirecting `ConciseView` output to a file?

Today, the most reliable word is `FullyQualifiedErrorID`, not error because that word may be used many times in custom messages. That gives me a distinct count of the errors and a quick Find Next to the next one.

I don't like that it's verbose, but `ErrorRecord: ` is a good candidate.

Reply:
> Interested how others feel about having ERROR or not having ERROR?

Error messages need to be searchable, so I want _something_ there. I'd vote for `Error:` over `ErrorRecord:` or anything more verbose than that. I can just use a regex search to match for text at the start of the line if I want to disambiguate from the word "error" appearing elsewhere in the output.

Reply:
Great conversation and thank you!  Rather than a static word that wasn't adding to the clarity of the error message, we instead populate the first word based on the error.  (see the rules here) In my testing, it has been much clearer in multiple cases.  What do you think?  
The error message will contain a prefix as described:

    + If the error is an Exception, it prefixes with Exception:.
    + If the error has InvocationInfo.MyCommand, it prefixes the command.
    + If the error has InvocationName, CategoryInfo.Category, or CategoryInfo.Reason,
the message will prefix these.
    + Only if none of those exist does it actually use Error:.

Reply:
Great! Would it be technically possible to include a hint whether it is a terminating or non-terminating error? This has always been intransparent and confusing to the user and has great practical value for error handling: terminating errors are not controlled via -ErrorAction whatsoever and will show even with -ErrorAction Ignore, whereas non-terminating errors require an -ErrorAction Stop to be able to handle them in try...catch. So knowing the type of error would provide a lot of clarity as to what can be done about the error, and how.

Reply:
@TobiasPSP: See https://github.com/PowerShell/PowerShell/pull/10115.

Reply:
Many thanks, excellent!

Reply:
> Great conversation and thank you!  Rather than a static word that wasn't adding to the clarity of the error message, we instead populate the first word based on the error.  (see the rules here) In my testing, it has been much clearer in multiple cases.  What do you think?

It may help with clarity but it doesn't help the search situation. Right now, I'm working on uploading 1,000+ files to a server. These are non-terminating errors so that it keeps uploading files but I may find 20 or so failed.. I go back to the review the logs, what do I search for? Each error could be a different type!

After thinking about it, I think `FullyQualifiedErrorId` should always be surfaced. It doesn't have to be prefixed, just needs to be in the error somewhere. Its what I use today to search and it allows you to search for specific ErrorIDs if needed.

Reply:
@dragonwolf83 does the open PR that @KirkMunro and @TobiasPSP are referencing help with your concerns? See PowerShell/PowerShell#10115

Reply:
No, I'm looking for all errors, regardless of type. You can have multiple terminating errors in a script depending on how you handle the errors.

I sanitized a real log file for an example below. It contains a mix of success and failed uploads. Imagine this being 100 files and say 22 failed. You want to quickly find the 22 errors.

If you search for `error`, you will get too many results because its used a lot in messages and in Write-Error itself. This means I will be hitting next so many times that I can easily miss the next error.

The only reliable way today is to search for `FullyQualifiedErrorId`. Regardless if you use `Throw` or `Write-Error`, it is always there.

In this log file, searching `error` results in 12 results and `FullyQualifiedErrorId` gives me 2 results.
```
Performing the operation "Upload" on target "C:\PowerShell\Test1.csv.zip".
Extract C:\PowerShell\Test1.csv.zip
Uploading C:\PowerShell\Test1.csv
C:\PowerShell\Test-Script.ps1: Send-Cmdlet: **Upload of file 'C:\PowerShell\Test1.csv' was successful, but error occurred while setting the permissions and/or timestamp.**
If the problem persists, turn off setting permissions or preserving timestamp. Alternatively you can turn on 'Ignore permission errors' option.
No such file or directory.
At line:1 char:1
+ .\Test-Script.ps1
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorExcep    tion
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorExceptio    n,Test-Script.ps1

Performing the operation "Upload" on target "C:\PowerShell\Test2.csv.zip".
Extract C:\PowerShell\Test2.csv.zip
Uploading C:\PowerShell\Test2.csv
Uploaded Completed
    
Performing the operation "Upload" on target "C:\PowerShell\Test3.csv.zip".
Extract C:\PowerShell\Test3.csv.zip
Uploading C:\PowerShell\Test3.csv
C:\PowerShell\Test-Script.ps1: Send-Cmdlet: **Upload of file 'C:\PowerShell\Test3.csv' was successful, but error occurred while setting the permissions and/or timestamp.**
If the problem persists, turn off setting permissions or preserving timestamp. Alternatively you can turn on 'Ignore permission errors' option.
No such file or directory.
At line:1 char:1
+ .\Test-Script.ps1
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorExcep    tion
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorExceptio    n,Test-Script.ps1


Performing the operation "Upload" on target "C:\PowerShell\Test4.csv.zip".
Extract C:\PowerShell\Test4.csv.zip
Uploading C:\PowerShell\Test4.csv
Uploaded Completed    
```

Now this happens to be the same error, but in this code, there can be a non-terminating error, a cmdlet error, a terminating error, an exception, etc. There needs to be a unifying, unique word that describes an `Error[0]` stream from regular output. This is why log files always prefix their output with `[INFO]`, `[DEBUG]`, `[ERROR]`. Easy to see and search. 

We don't have to surface `FullyQualifiedErrorId`. We just need that one word that will be consistently used across all error types.

Reply:
To support searchability of transcription logs, we may want to log with NormalView so that FullyQualifiedErrorId is there.

Reply:
Yeah, I get the use case here, though, and the complaint is pretty valid. I'm not transcripting in CI/CD pipelines, it's just a straight log of stdout....lots of tools use things like `[E]` to denote the output stream (Android's `adb logcat` comes to mind originally, but it's a fairly standard practice). I guess the old formatting technically fulfilled this with something consistent like "FullyQualifiedErrorId" in the output....

But then again, this isn't done by Bash's stdout at all, and writing (and capturing in your own logic) the exit code is the accepted norm there. And nothing's stopping you from switching back to the old error view in your `$ErrorView`. I'll bring it up as a discussion point today.

Reply:
@PowerShell/powershell-committee talked about this. Given the amount of original feedback pushing back against polluting with an additional "Error:", we're okay with the workaround of using a different `$ErrorView` if you want to be able to globally Ctrl+F in logs for errors. 

If this proves to be a larger problem than we think, we can revisit and change this (given it's formatting).